### ISSUE TITLE
Some fields need pruning control for CRD

### ISSUE THREAD

**Issue by snuggie12** [jaegertracing/jaeger-operator#1314]
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#controlling-pruning

When building a new cluster I ran into an issue where any part of the spec using the `options` field would get emptied (i.e. `options: {}`) after applying the yaml.

For some reason I noticed on a different cluster the CRD for `jaegers.jaegertracing.io` had a field `spec.preserveUnknownFields`. I altered my CRD in the new cluster to have this and the options stuck this time.

This added a bunch of empty fields. I suspect the proper fix is to individually put the `x-kubernetes-preserve-unknown-fields` in place and then fields I did not apply will not automatically get added.

**Comment by jpkrohling** [comment]
Would you mind explaining what's the actual issue this is causing? Or is it only about adding an empty object when the property isn't set?

**Comment by snuggie12** [comment]
> Would you mind explaining what's the actual issue this is causing? Or is it only about adding an empty object when the property isn't set?

I'm specifying Kafka options on the collector, for instance, but the object gets emptied out. Since options.kafka is not in the open api spec it gets pruned and my options turns to `options: {}`.

**Comment by jpkrohling** [comment]
I think I know what you mean, but could you please provide a CR that would make it clear what the problem is?

**Comment by snuggie12** [comment]
Sure thing

```
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger-two
  namespace: monitoring
spec:
  collector:
    maxReplicas: 3
    options:
      kafka:
        producer:
          brokers: 'jaeger-kafka-brokers.monitoring:9092'
          topic: jaeger-spans
...
```

After a `kubectl apply` it becomes:

```
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger-two
  namespace: monitoring
spec:
  collector:
    maxReplicas: 3
    options: {}
...
```

**Comment by snuggie12** [comment]
Creating your CRD to look like this:

```
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    controller-gen.kubebuilder.io/version: v0.2.4
  name: jaegers.jaegertracing.io
spec:
  conversion:
    strategy: None
  group: jaegertracing.io
  names:
    kind: Jaeger
    listKind: JaegerList
    plural: jaegers
    singular: jaeger
  preserveUnknownFields: true # added field
...
```

Will keep my kafka options but you are correct that a bunch of other undefined fields get pointlessly created (such as `spec.agent.config` in this example). I haven't looked into this too much, but I think if you leave `spec.preserveUnknownFields` undefined but you use `x-kubernetes-preserve-unknown-fields` on the individual fields it might look correct? Not really sure, just speculating.

**Comment by jpkrohling** [comment]
@rubenvp8510, is this something you can take a look at?

**Comment by rubenvp8510** [comment]
How did you create the CDR? 

I tried to reproduce this with kubernetes 1.16 (tested with 1.19 also) and I'm not able to reproduce it when I created the CRD using  `deploy/crds/jaegertracing.io_jaegers_crd.yaml` of `make crd`

The only way I was able to reproduce this is with the following steps:

1) `make crd `or` kubectl create -f  deploy/crds/jaegertracing.io_jaegers_crd.yaml`
2) `kubectl get crd jaegers.jaegertracing.io -o yaml > crd.yaml`
3)  Edit `crd.yaml `and remove `spec.preserveUnknownFields`
4) `kubectl delete jaegers.jaegertracing.io`
5) `kubectl create -f crd.yaml`

Otherwise I always get the spec.preserveUnknownFields=true on the final CRD. According to the kubernetes documentation

> CRDs converted from apiextensions.k8s.io/v1beta1 to apiextensions.k8s.io/v1 might lack structural schemas, and spec.preserveUnknownFields might be true. (1)

The current CRD uses `apiextensions.k8s.io/v1beta1` isn't it? and kubernetes convert it to  `apiextensions.k8s.io/v1`

1) https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#field-pruning

**Comment by snuggie12** [comment]
Yes, you are correct. This was a new cluster and all CRDs are bootstrapped prior to running ArgoCD. When a fellow teammate was building this new cluster they made sure the CRDs we have on a cluster was updated by essentially doing a `kubectl get crds -o yaml > crds.yaml`. As this was done on a 1.16+ cluster they were outputted with `apiVersion: apiextensions.k8s.io/v1`.

What's worse is that I wasn't told we ran into issues with this process because of the `preserveUnknownFields: true` and those fields were actually removed and later re-added.

Technically I could see this issue being closed, but I do think at some point you'll need to update your CRDs to have a complete spec or have specific fields marked with `x-kubernetes-preserve-unknown-fields`. I'll leave that up to you all whether that's this issue or a future one.
